torchx/lib/torchx/backend.ex

```diff
@@ -0,0 +1,219 @@
+  end
+
+  def backend_transfer(tensor, Torchx.Backend, _opts) do
+    tensor
+  end
+
+  def backend_transfer(tensor, backend, opts) do
+    backend.from_binary(tensor, to_blob(tensor), opts)
+  end
+
+  @impl true
+  def from_binary(%{type: type, shape: shape} = out, binary, _opts) do
+    NIF.from_blob(binary, shape, torch_type(type)) |> from_ref(out)
+  end
+
+  ## Shape
+
+  @impl true
+  def reshape(out, %{data: %{ref: ref}} = tensor, shape),
+    do: NIF.reshape(ref, shape) |> from_ref(tensor)
+
+  @impl true
+  def as_type(%{type: type} = out, %{data: %{ref: ref}} = tensor),
+    do: NIF.to_type(ref, torch_type(type)) |> from_ref(out)
+
+  @impl true
+  def squeeze(out, tensor, _axes) do
+    NIF.squeeze(tensor.data.ref) |> from_ref(out)
+  end
+
+  @impl true
+  def transpose(out, tensor, axes), do: IO.inspect(axes)
+  # do: NIF.transpose(tensor, dim0, dim1) |> from_ref(out)
+
+  ## Ops
+
+  @impl true
+  def add(out, left, right) do
+    NIF.add(left.data.ref, right.data.ref) |> from_ref(out)
+  end
+
+  @impl true
+  def dot(
+        out,
+        %{type: t1, data: %{ref: left_ref}} = left,
+        _axes1,
+        %{type: t2, data: %{ref: right_ref}} = right,
+        _axes2
+      ) do
+    NIF.dot(left_ref, right_ref) |> from_ref(out)
+  end
+
+  @impl true
+  def cholesky(%{type: output_type, shape: {rows, cols}} = out, tensor) do
+    NIF.cholesky(tensor.data.ref) |> from_ref(out)
+  end
+
+  @impl true
+  def qr(
+        {q_holder, r_holder},
+        tensor,
+        opts
+      ),
+      do: NIF.qr(tensor.data.ref, opts[:mode] == :reduced) |> from_ref({q_holder, r_holder})
+
+  @impl true
+  def inspect(%{type: {_, elem_size}} = tensor, inspect_opts) do
+    limit = if(inspect_opts.limit == :infinity, do: nil, else: inspect_opts.limit + 1)
+
+    if on_cpu?(tensor) do
+      byte_size = nbytes(tensor)
+      byte_limit = limit && limit * div(elem_size, 8)
+
+      if(min(byte_limit, byte_size) > @big_tensor_threshold) do
+        raise "Tensor is too big (#{byte_size} bytes) for to_binary/1 operation." <>
+                "You must first transfer the tensor to Elixir by calling Nx.backend_transfer/1"
+      else
+        binary = to_blob(tensor, limit)
+        Nx.Backend.inspect(tensor, binary, inspect_opts)
+      end
+    else
+      raise "Tensor is located on #{device(tensor)} device, so direct to_binary/1 operation is not supported. " <>
+              "You must first transfer the tensor to Elixir by calling Nx.backend_transfer/1"
+    end
+  end
+
+  defp unwrap!({:ok, result}), do: result
+
+  defp unwrap!({:error, error}),
+    do: raise(RuntimeError, "PyTorch: " <> to_string(error))
+
+  defp from_ref({:ok, ref}, t), do: from_ref(ref, t)
+  defp from_ref({:error, _error} = input, _t), do: unwrap!(input)
+  defp from_ref({ref1, ref2}, {t1, t2}), do: {from_ref(ref1, t1), from_ref(ref2, t2)}
+  defp from_ref([ref | list], t), do: [from_ref(ref, t) | from_ref(list, t)]
+  defp from_ref([], _t), do: []
```

Comment:
I would have those as distinct functions, so the code is clearer: `from_pair_ref` when working with tuples, `from_list_ref` when working with lists, and so on. This would allow us to assert what is the value we are expecting from the NIF.

Examples:

    defp from_ref(maybe_ref, t), do: maybe_ref |> unwrap!() |> to_tensor(t)

    defp from_pair_ref(maybe_ref, t) do
      {left, right} = unwrap!(maybe_ref)
      {to_tensor(left, t), to_tensor(right, t)}
    end

    defp to_tensor(ref, t), do: %{t | data: %__MODULE__{ref: ref}}